src/OrchardCore/OrchardCore/Shell/Builders/Extensions/ServiceProviderExtensions.cs

```diff
@@ -75,65 +76,44 @@ public static IServiceCollection CreateChildContainer(this IServiceProvider serv
                     }
                 }
 
-                // If all services of the same type are not singletons.
-                else if (services.All(s => s.Lifetime != ServiceLifetime.Singleton))
-                {
-                    // We don't need to resolve them.
-                    foreach (var service in services)
-                    {
-                        clonedCollection.Add(service);
-                    }
-                }
-
-                // If all services of the same type are singletons.
-                else if (services.All(s => s.Lifetime == ServiceLifetime.Singleton))
+                // If services of the same type have at least one singleton.
+                else if (services.Any(s => s.Lifetime == ServiceLifetime.Singleton))
                 {
                     // We can resolve them from the main container.
-                    var instances = services.Key.ServiceKey is not null
-                        ? serviceProvider.GetKeyedServices(services.Key.ServiceType, services.Key.ServiceKey)
-                        : serviceProvider.GetServices(services.Key.ServiceType);
-
-                    for (var i = 0; i < services.Count(); i++)
-                    {
-                        var instance = instances.ElementAt(i);
-                        if (instance is null)
-                        {
-                            continue;
-                        }
-
-                        clonedCollection.CloneSingleton(services.ElementAt(i), instance);
-                    }
-                }
-
-                // If singletons and scoped services are mixed.
-                else
-                {
-                    // We need a service scope to resolve them.
-                    using var scope = serviceProvider.CreateScope();
```

Comment:
Under https://github.com/OrchardCMS/OrchardCore/issues/15361#issuecomment-2137342885 you said that: 

> If I understand this code correctly, it is an optimization if there is a single service only. In that case, if the result is not disposable, creating the singletons can be deferred until a tenant resolves them. Otherwise all services must be created immediately.
> So I think this should be kept, but the code is hard to understand. Maybe we can simplify it little bit.

I'm not sure I'm following: here you did remove this line in the end. I don't think it makes sense (while it's not obvious to me is because `CreateScope()` may actually do things that affect services in its parent scope), but you said it's somehow an optimization. Is it, then, and if yes, how does it work? Why don't we need it now?

Reply:
Sorry for the confusion, the optimization I was referring to was the different behavior when a between a single and multiple services (as from the original bug report). That was unrelated to the removal of `CreateScope` and is kept the same.

Reply:
Creating the scope here doesn't have any effect as far as I know, at least as long as it isn't used. But the actual intent may have been to use the scope, as any scoped application services are kept by the parent service provider alive forever. But I was explicitly keeping the functionality the same for now.